ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -0,0 +1,9955 @@
+|					jnz >3
+||				} else {
+|					jnz >4
+||				}
+||			}
+||			if (safe) {
+|				// ZVAL_NULL(cv);
+|				SET_ZVAL_TYPE_INFO addr, IS_NULL
+||			}
+|			// zval_dtor_func(r);
+|			ZVAL_DTOR_FUNC op_info, opline
+||			if (gc && RC_MAY_BE_N(op_info) && ((op_info) & (MAY_BE_REF|MAY_BE_ARRAY|MAY_BE_OBJECT))) {
+|				jmp >4
+||			}
+|3:
+||		}
+||		if (gc && RC_MAY_BE_N(op_info) && ((op_info) & (MAY_BE_REF|MAY_BE_ARRAY|MAY_BE_OBJECT))) {
+||			if ((op_info) & MAY_BE_REF) {
+||				zend_jit_addr ref_addr = ZEND_ADDR_MEM_ZVAL(ZREG_FCARG1a, offsetof(zend_reference, val));
+|				IF_NOT_ZVAL_TYPE addr, IS_REFERENCE, >1
+|				IF_NOT_ZVAL_REFCOUNTED ref_addr, >4
+|				GET_ZVAL_PTR FCARG1a, ref_addr
+|1:
+||			}
+|			IF_GC_MAY_NOT_LEAK FCARG1a, eax, >4
+|			// gc_possible_root(Z_COUNTED_P(z))
+|			EXT_CALL gc_possible_root, r0
+||			if (cold && ((op_info) & ((MAY_BE_ANY|MAY_BE_UNDEF)-(MAY_BE_OBJECT|MAY_BE_RESOURCE)))) {
+|				jmp >4
+|.code
+||			}
+||		} else if (cold && ((op_info) & ((MAY_BE_ANY|MAY_BE_UNDEF)-(MAY_BE_OBJECT|MAY_BE_RESOURCE)))) {
+|			jmp >4
+|.code
+||		}
+|4:
+||	}
+|.endmacro
+
+|.macro FREE_OP, op_type, op, op_info, cold, op_array, opline
+||	if (op_type & (IS_VAR|IS_TMP_VAR)) {
+|		ZVAL_PTR_DTOR ZEND_ADDR_MEM_ZVAL(ZREG_FP, op.var), op_info, 0, cold, 0, opline
+||	}
+|.endmacro
+
+|.macro SEPARATE_ZVAL_NOREF, addr, op_info, cold
+||	if ((op_info & MAY_BE_ARRAY) && RC_MAY_BE_N(op_info)) {
+||		if (cold) {
+|			IF_ZVAL_TYPE addr, IS_ARRAY, >1
+|.cold_code
+|1:
+||		} else {
+|			IF_NOT_ZVAL_TYPE addr, IS_ARRAY, >2
+||		}
+|		GET_ZVAL_PTR r0, addr
+||		if (RC_MAY_BE_1(op_info)) {
+|			cmp dword [r0], 1 // if (GC_REFCOUNTED() > 1)
+|			jbe >2
+||		}
+|		IF_NOT_ZVAL_FLAGS addr, IS_TYPE_REFCOUNTED, >1
+|		GC_DELREF r0
+|1:
+||		if (Z_REG(addr) == ZREG_FCARG1a) {
+|			mov aword [r4], FCARG1a // save
+||		} else {
+|			LOAD_ZVAL_ADDR FCARG1a, addr
+||		}
+|		EXT_CALL zval_copy_ctor_func, r0
+||		if (Z_REG(addr) == ZREG_FCARG1a) {
+|			mov FCARG1a, aword [r4] // restore
+||		}
+||		if (cold) {
+|			jmp >2
+|.code
+||		}
+|2:
+||	}
+|.endmacro
+
+|.macro SEPARATE_ARRAY, addr, op_info, cold
+||	if (RC_MAY_BE_N(op_info)) {
+||		zend_reg tmp_reg;
+||
+||		tmp_reg = (Z_REG(addr) == ZREG_FCARG1a) ? ZREG_R0 : ZREG_FCARG1a;
+|		GET_ZVAL_LVAL tmp_reg, addr
+||		if (RC_MAY_BE_1(op_info)) {
+|			cmp dword [Ra(tmp_reg)], 1 // if (GC_REFCOUNTED() > 1)
+||			if (cold) {
+|				ja >1
+|.cold_code
+|1:
+||			} else {
+|				jbe >2
+||			}
+||		}
+|		IF_NOT_ZVAL_FLAGS addr, IS_TYPE_REFCOUNTED, >1
+|		GC_DELREF Ra(tmp_reg)
+|1:
+||		if (Z_REG(addr) == ZREG_FCARG1a) {
+|			mov aword [r4], FCARG1a // save
+||		} else {
+|			LOAD_ZVAL_ADDR FCARG1a, addr
+||		}
+|		EXT_CALL zval_copy_ctor_func, r0
+||		if (Z_REG(addr) == ZREG_FCARG1a) {
+|			mov FCARG1a, aword [r4] // restore
+||		}
+||		if (RC_MAY_BE_1(op_info)) {
+||			if (cold) {
+|				jmp >2
+|.code
+||			}
+||		}
+|2:
+||	}
+|		GET_ZVAL_LVAL ZREG_FCARG1a, addr
```

Comment:
Worth pointing out this side-effect in a comment above the macro.

Reply:
Probably, it makes sense to add uniform description for all macros.